File: HelloOCRuntime/Utils/OakProperty.m
//
//  OakProperty.m
//  HelloOCRuntime
//
//  Created by d on 2023/07/22.
//

#import "OakProperty.h"

@import Foundation;
@import ObjectiveC;

@implementation OakProperty

- (instancetype)initWithName:(NSString *)name attributes:(NSString *)attributes {
  self = [super init];
  if (self) {
    self.name = name;
    self.attributes = attributes;
  }
  return self;
}

- (NSString *)description {
  return [[NSString alloc] initWithFormat:@"<%p: name: %@, attributes: %@>", self, self.name, self.attributes];
}

@end


NSArray<OakProperty *> * oak_propertics(Class klass) {
  NSMutableArray<OakProperty *>* oak_properties = [NSMutableArray array];
  unsigned int count_of_properties = 0;
  objc_property_t* properties = class_copyPropertyList(klass, &count_of_properties);
  for (NSUInteger i = 0; i < count_of_properties; i++) {
    objc_property_t property = properties[i];
    const char* property_name = property_getName(property);
    const char* property_attributes = property_getAttributes(property);
    NSString* name = [[NSString alloc] initWithUTF8String:property_name];
    NSString* attributes = [[NSString alloc] initWithUTF8String:property_attributes];
    [oak_properties addObject:[[OakProperty alloc] initWithName:name attributes:attributes]];
  }
  free(properties);
  return [oak_properties copy];
}

File: HelloOCRuntime/Utils/OakProperty.swift
//
//  OakProperty.swift
//  HelloOCRuntime
//
//  Created by d on 2023/07/23.
//

import Foundation
import ObjectiveC

@frozen
public struct OakProperty {
  let name: String
  let attributes: String
  let typeEncoding: String
  let isReferenceType: Bool
  let ivarName: String
}

extension OakProperty: CustomStringConvertible {
  public var description: String {
"""
{
           name: \(name)
   typeEncoding: \(typeEncoding)
isReferenceType: \(isReferenceType)
       ivarName: \(ivarName)
}
"""
  }
}

// MARK: - oak_propertics

func oak_propertics(_ klass: AnyClass?) -> [OakProperty]? {
  var count_of_properties: UInt32 = 0
  var oak_properties = [OakProperty]()
  let properties = class_copyPropertyList(klass, &count_of_properties)
  defer { properties?.deallocate() }
  
  for i in 0..<Int(count_of_properties) {
    guard let property = properties?[i] else {
      continue
    }
    
    var attributes = "no attributes"
    if let s = property_getAttributes(property) {
      attributes = String(cString: s)
    }
    
    var count: UInt32 = 0
    let attributeList = property_copyAttributeList(property, &count)
    var typeEncoding = ""
    var isReferenceType = false
    var ivarName = ""
    for i in 0..<Int(count) {
      guard let attribute = attributeList?[i] else {
        continue
      }
      let name = String(cString: attribute.name)
      if name == "T" {
        let value = String(cString: attribute.value).trimmingCharacters(in: CharacterSet(charactersIn: "@\""))
        typeEncoding = value
      } else if name == "&" {
        isReferenceType = true
      } else if name == "V" {
        ivarName = String(cString: attribute.value)
      }

    }
    
    oak_properties.append(OakProperty(name: String(cString: property_getName(property)),
                                      attributes: attributes,
                                      typeEncoding: typeEncoding,
                                      isReferenceType: isReferenceType,
                                      ivarName: ivarName))
  }
  return oak_properties
}
